src/cuda/cuda_sampling.cu

```diff
@@ -537,32 +540,534 @@ void LaunchSort(SamplingData* data, cudaStream_t stream, float* scores_in, float
+    const int kBlockSize = 256;
+    const int max_k = 64;
+
+    float* intermediate_scores = data->scores_buffer.get();
+    int* intermediate_indices = data->indices_in.get();
+
+    dim3 grid_stage1(num_partitions, batch_size);
+    dim3 block_stage1(kBlockSize);
+    FindBlockTopK<max_k, kBlockSize><<<grid_stage1, block_stage1, 0, stream>>>(
+        scores_in, intermediate_indices, intermediate_scores, vocab_size, num_partitions);
+
+    int num_intermediate_results = num_partitions * max_k;
+    dim3 grid_stage2(batch_size);
+    dim3 block_stage2(kBlockSize);
+    ReduceFinalTopKAndSoftmax<max_k, kBlockSize><<<grid_stage2, block_stage2, 0, stream>>>(
+        indices_out, scores_out, intermediate_indices, intermediate_scores,
+        num_intermediate_results, k, temperature);
+}
+
+void RunTopKViaMapReduceShared(SamplingData* data, cudaStream_t stream, float* scores_in, float* scores_out, int* indices_out, int vocab_size, int batch_size, int k, float temperature, int num_partitions) {
+    const int kBlockSize = 256;
+    const int max_k = 64;
+
+    float* intermediate_scores = data->scores_buffer.get();
+    int* intermediate_indices = data->indices_in.get();
+
+    dim3 grid_stage1(num_partitions, batch_size);
+    dim3 block_stage1(kBlockSize);
+    FindBlockTopK<max_k, kBlockSize><<<grid_stage1, block_stage1, 0, stream>>>(
+        scores_in, intermediate_indices, intermediate_scores, vocab_size, num_partitions);
+
+    int num_intermediate_results = num_partitions * max_k;
+    size_t shared_mem_size = num_intermediate_results * (sizeof(float) + sizeof(int));
+
+    dim3 grid_stage2(batch_size);
+    dim3 block_stage2(kBlockSize);
+    ReduceFinalTopKAndSoftmax_SharedMemory<max_k, kBlockSize><<<grid_stage2, block_stage2, shared_mem_size, stream>>>(
+        indices_out, scores_out, intermediate_indices, intermediate_scores,
+        num_intermediate_results, k, temperature);
+}
+
+void RunTopKViaSingleKernelMapReduce(SamplingData* data, cudaStream_t stream, float* scores_in, float* scores_out, int* indices_out, int vocab_size, int batch_size, int k, float temperature, int num_partitions) {
+    const int kBlockSize = 256;
+    const int max_k = 64;
+
+    // This algorithm is specialized for batch_size = 1
+    if (batch_size != 1) {
+        // Fallback or assert
+        RunTopKViaMapReduce(data, stream, scores_in, scores_out, indices_out, vocab_size, batch_size, k, temperature, num_partitions);
+        return;
+    }
+
+    float* intermediate_scores = data->scores_buffer.get();
+    int* intermediate_indices = data->indices_in.get();
+
+    cudaMemsetAsync(data->sync_counter.get(), 0, sizeof(int), stream);
+
+    dim3 grid(num_partitions);
+    dim3 block(kBlockSize);
+    SingleKernelMapReduceTopKAndSoftmax<max_k, kBlockSize><<<grid, block, 0, stream>>>(
```

Comment:
Technically, we need to "reset" the sync counter (either via cudaMemsetAsync outside the kernel or inside the kernel once we cross the synchronization step) - its cost muist be accounted for while measuring the E2E latency of this approach